Aircraft module code refactoring

charliefoxtwo opened the pull request:
This adds typing for the core MemoryMap and Documentation tables, and migrates the A-4E-C and P-51D to table-based modules.

For the most part, module definition is unchanged. The generated json files are identical, meaning address assignment logic has been preserved and this will have no impact to clients.

This doesn't _fully_ remove global variables as there are still a few that are more deeply embedded, so I've kept that functionality in Protocol.lua and added some wrapper functions to maintain that functionality. I'd like to remove them in the future.

I've left lots of documentation marked as TODO as I just don't know the full reasoning for the existence of some of these items that appear in the json.

Unit tests have been added for the control definition functionality that has been migrated thus far. As we migrate more, it will be imperative to continue adding unit tests to verify functionality.

jdahlblom commented:
Tested with TF-51D and A-10C II and it seems to work. Got all data from the export and comms back seems to work.

jdahlblom commented:
[P-51D.zip](https://github.com/DCSFlightpanels/dcs-bios/files/12569569/P-51D.zip)

And I added :
```
local P_51D = require "lib.module.P-51D"
BIOS.protocol.writeNewModule(P_51D)```

charliefoxtwo commented:
> [P-51D.zip](https://github.com/DCSFlightpanels/dcs-bios/files/12569569/P-51D.zip)
> 
> And I added :
> 
> ```
> local P_51D = require "lib.module.P-51D"
> BIOS.protocol.writeNewModule(P_51D)```
> ```

Thanks for reporting this! I've found and fixed the bug. New modules should properly receive and export data now!

As an aside, in testing this I also uncovered a bug with the P-51's configuration (specifically `RKT_COUNT_CON` is incorrect), but we can fix that later as that bug existed in prior versions of bios.

WarLord211 commented:
what is with the RKT_COUNT_CON ?

jdahlblom commented:
> > [P-51D.zip](https://github.com/DCSFlightpanels/dcs-bios/files/12569569/P-51D.zip)
> > And I added :
> > ```
> > local P_51D = require "lib.module.P-51D"
> > BIOS.protocol.writeNewModule(P_51D)```
> > ```
> 
> Thanks for reporting this! I've found and fixed the bug. New modules should properly receive and export data now!
> 
> As an aside, in testing this I also uncovered a bug with the P-51's configuration (specifically `RKT_COUNT_CON` is incorrect), but we can fix that later as that bug existed in prior versions of bios.

What exactly did you do? I also had the P_51D module declaration in BIOS.lua.

charliefoxtwo commented:
> What exactly did you do? I also had the P_51D module declaration in BIOS.lua.

One-liner :) https://github.com/DCSFlightpanels/dcs-bios/pull/197/commits/7f889cd49a3cd99af5966db8c2a7d3cc6afc1c6e

charliefoxtwo commented:
I've also gotten unit tests working locally! They're not in this pr - I'll include them in a future one - but currently they validate that IDs aren't duplicated and match the defined format (only uppercase letters, numbers, and underscores) for any given module. I've also started writing tests to validate that some of the behavior for the different control definition functions is correct.
